File: training/CombatModelVerifier/DecisionParity.cs
internal sealed record DecisionParityResult(
    int SampleCount,
    double Top1AgreementRate,
    double Top3SetAgreementRate,
    double RankingAgreement,
    bool TieBreakDeterministic,
    int NearTieSampleCount,
    int TieSampleCount,
    long DiscordantPairs,
    long ComparedPairs)
{
    public bool Passed => Top1AgreementRate == 1.0 && Top3SetAgreementRate == 1.0 &&
                          RankingAgreement == 1.0 && TieBreakDeterministic;
}

internal static class DecisionParity
{
    public const double TieTolerance = 1e-4;
    public const double NearTieThreshold = 1e-3;

    public static int[] CanonicalRanking(IReadOnlyList<double> scores, IReadOnlyList<string> actionIds,
        IReadOnlyList<bool> legal, double tieTolerance = TieTolerance)
    {
        var ordered = Enumerable.Range(0, scores.Count).Where(index => legal[index])
            .OrderByDescending(index => scores[index])
            .ThenBy(index => actionIds[index], StringComparer.Ordinal)
            .ToArray();
        var result = new List<int>(ordered.Length);
        var cursor = 0;
        while (cursor < ordered.Length)
        {
            var anchor = scores[ordered[cursor]];
            var end = cursor + 1;
            while (end < ordered.Length && anchor - scores[ordered[end]] < tieTolerance) end++;
            result.AddRange(ordered[cursor..end].OrderBy(index => actionIds[index], StringComparer.Ordinal));
            cursor = end;
        }
        return result.ToArray();
    }

    public static DecisionParityResult Compare(double[,] reference, double[,] candidate,
        IReadOnlyList<string[]> actionIds, bool[,] legalMask, double tieTolerance = TieTolerance)
    {
        var samples = reference.GetLength(0);
        var actions = reference.GetLength(1);
        if (candidate.GetLength(0) != samples || candidate.GetLength(1) != actions ||
            legalMask.GetLength(0) != samples || legalMask.GetLength(1) != actions || actionIds.Count != samples)
            throw new InvalidOperationException("decision parity shapes differ");
        var top1 = 0;
        var top3 = 0;
        var nearTies = 0;
        var ties = 0;
        long concordant = 0;
        long discordant = 0;
        var deterministic = true;
        for (var sample = 0; sample < samples; sample++)
        {
            var ids = actionIds[sample];
            var legalCount = Enumerable.Range(0, actions).Count(index => legalMask[sample, index]);
            if (ids.Length != legalCount)
                throw new InvalidOperationException($"sample {sample} action ids {ids.Length} != legal actions {legalCount}");
            var paddedIds = ids.Concat(Enumerable.Range(ids.Length, actions - ids.Length).Select(index => $"<PAD:{index}>")).ToArray();
            var legal = Enumerable.Range(0, actions).Select(index => legalMask[sample, index]).ToArray();
            var referenceScores = Enumerable.Range(0, actions).Select(index => reference[sample, index]).ToArray();
            var candidateScores = Enumerable.Range(0, actions).Select(index => candidate[sample, index]).ToArray();
            var expected = CanonicalRanking(referenceScores, paddedIds, legal, tieTolerance);
            var actual = CanonicalRanking(candidateScores, paddedIds, legal, tieTolerance);
            deterministic &= expected.SequenceEqual(CanonicalRanking(referenceScores, paddedIds, legal, tieTolerance));
            deterministic &= actual.SequenceEqual(CanonicalRanking(candidateScores, paddedIds, legal, tieTolerance));
            if (expected.Length > 0 && actual.Length > 0 && expected[0] == actual[0]) top1++;
            if (expected.Take(3).ToHashSet().SetEquals(actual.Take(3))) top3++;
            var actualPositions = actual.Select((action, position) => (action, position)).ToDictionary(item => item.action, item => item.position);
            for (var left = 0; left < expected.Length; left++)
            for (var right = left + 1; right < expected.Length; right++)
            {
                if (actualPositions[expected[left]] < actualPositions[expected[right]]) concordant++;
                else discordant++;
            }
            var legalScores = expected.Select(index => referenceScores[index]).OrderByDescending(value => value).ToArray();
            if (legalScores.Length >= 2)
            {
                var margin = legalScores[0] - legalScores[1];
                if (margin < NearTieThreshold) nearTies++;
                if (margin < tieTolerance) ties++;
            }
        }
        var pairs = concordant + discordant;
        return new DecisionParityResult(samples, (double)top1 / Math.Max(samples, 1),
            (double)top3 / Math.Max(samples, 1), (double)(concordant - discordant) / Math.Max(pairs, 1),
            deterministic, nearTies, ties, discordant, pairs);
    }
}

File: training/combat_model/decision_parity.py
"""Decision-equivalence metrics for PyTorch/ONNX combat policy logits."""

from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Sequence

import numpy as np


TIE_TOLERANCE = 1.0e-4
NEAR_TIE_THRESHOLD = 1.0e-3
NUMERIC_DIAGNOSTIC_THRESHOLD = 1.0e-3
LEGACY_ABSOLUTE_TOLERANCE = 1.0e-5


def canonical_ranking(
    scores: Sequence[float],
    action_ids: Sequence[str],
    legal: Sequence[bool],
    tie_tolerance: float = TIE_TOLERANCE,
) -> list[int]:
    """Rank legal actions, resolving an anchor-relative tie group by ActionId."""
    ordered = sorted(
        (index for index, is_legal in enumerate(legal) if is_legal),
        key=lambda index: (-float(scores[index]), action_ids[index]),
    )
    result: list[int] = []
    cursor = 0
    while cursor < len(ordered):
        anchor = float(scores[ordered[cursor]])
        end = cursor + 1
        while end < len(ordered) and anchor - float(scores[ordered[end]]) < tie_tolerance:
            end += 1
        result.extend(sorted(ordered[cursor:end], key=lambda index: action_ids[index]))
        cursor = end
    return result


@dataclass(frozen=True)
class DecisionParityResult:
    sample_count: int
    top1_agreement_rate: float
    top3_set_agreement_rate: float
    ranking_agreement: float
    tie_break_deterministic: bool
    near_tie_sample_count: int
    tie_sample_count: int
    discordant_pairs: int
    compared_pairs: int

    @property
    def passed(self) -> bool:
        return (
            self.top1_agreement_rate == 1.0
            and self.top3_set_agreement_rate == 1.0
            and self.ranking_agreement == 1.0
            and self.tie_break_deterministic
        )

    def to_dict(self) -> dict[str, object]:
        return {**asdict(self), "verdict": "pass" if self.passed else "fail"}


def compare_decisions(
    reference_logits: np.ndarray,
    candidate_logits: np.ndarray,
    action_ids: Sequence[Sequence[str]],
    legal_mask: np.ndarray,
    tie_tolerance: float = TIE_TOLERANCE,
) -> DecisionParityResult:
    if reference_logits.shape != candidate_logits.shape or reference_logits.shape != legal_mask.shape:
        raise ValueError("reference, candidate, and legal-mask shapes must match")
    if reference_logits.shape[0] != len(action_ids):
        raise ValueError("action_ids sample count must match logits")
    top1 = top3 = 0
    concordant = discordant = 0
    deterministic = True
    near_ties = ties = 0
    for sample_index in range(reference_logits.shape[0]):
        legal_count = int(np.count_nonzero(legal_mask[sample_index]))
        ids = list(action_ids[sample_index])
        if len(ids) != legal_count:
            raise ValueError(f"sample {sample_index} action_ids={len(ids)}, legal={legal_count}")
        padded_ids = ids + [f"<PAD:{index}>" for index in range(len(ids), reference_logits.shape[1])]
        legal = legal_mask[sample_index].tolist()
        reference = canonical_ranking(reference_logits[sample_index], padded_ids, legal, tie_tolerance)
        candidate = canonical_ranking(candidate_logits[sample_index], padded_ids, legal, tie_tolerance)
        deterministic &= reference == canonical_ranking(reference_logits[sample_index], padded_ids, legal, tie_tolerance)
        deterministic &= candidate == canonical_ranking(candidate_logits[sample_index], padded_ids, legal, tie_tolerance)
        if reference and candidate and reference[0] == candidate[0]:
            top1 += 1
        if set(reference[:3]) == set(candidate[:3]):
            top3 += 1
        reference_positions = {action: position for position, action in enumerate(reference)}
        candidate_positions = {action: position for position, action in enumerate(candidate)}
        for left in range(len(reference)):
            for right in range(left + 1, len(reference)):
                first, second = reference[left], reference[right]
                if (reference_positions[first] - reference_positions[second]) * (candidate_positions[first] - candidate_positions[second]) > 0:
                    concordant += 1
                else:
                    discordant += 1
        legal_scores = sorted((float(reference_logits[sample_index, index]) for index in reference), reverse=True)
        if len(legal_scores) >= 2:
            margin = legal_scores[0] - legal_scores[1]
            near_ties += int(margin < NEAR_TIE_THRESHOLD)
            ties += int(margin < tie_tolerance)
    samples = reference_logits.shape[0]
    pairs = concordant + discordant
    return DecisionParityResult(
        sample_count=samples,
        top1_agreement_rate=top1 / max(samples, 1),
        top3_set_agreement_rate=top3 / max(samples, 1),
        ranking_agreement=(concordant - discordant) / max(pairs, 1),
        tie_break_deterministic=deterministic,
        near_tie_sample_count=near_ties,
        tie_sample_count=ties,
        discordant_pairs=discordant,
        compared_pairs=pairs,
    )


def inject_near_tie_perturbation(
    candidate_logits: np.ndarray,
    reference_logits: np.ndarray,
    action_ids: Sequence[Sequence[str]],
    legal_mask: np.ndarray,
    amount: float,
) -> tuple[np.ndarray, int, int]:
    """Add ``amount`` to the runner-up of the first near-tie sample."""
    modified = candidate_logits.copy()
    for sample_index in range(reference_logits.shape[0]):
        count = int(np.count_nonzero(legal_mask[sample_index]))
        ids = list(action_ids[sample_index])
        legal = legal_mask[sample_index, :count].tolist()
        ranking = canonical_ranking(reference_logits[sample_index, :count], ids, legal)
        if len(ranking) < 2:
            continue
        margin = float(reference_logits[sample_index, ranking[0]] - reference_logits[sample_index, ranking[1]])
        if margin < NEAR_TIE_THRESHOLD:
            modified[sample_index, ranking[1]] += amount
            return modified, sample_index, ranking[1]
    raise ValueError("fixture contains no near-tie sample for perturbation test")
